tools/mason/MasonBuild.chpl

```diff
@@ -0,0 +1,174 @@
+
+use TOML;
+use Spawn;
+use FileSystem;
+use MasonUtils;
+use MasonHelp;
+
+
+
+proc masonBuild(args) {
+  var show = false;
+  var release = false;
+  var compopts: [1..0] string;
+  if args.size > 2 {
+    for arg in args[2..] {
+      if arg == '-h' || arg == '--help' {
+	masonBuildHelp();
+	exit();
+      }
+      else if arg == '--release' {
+	release = true;
+      }
+      else if arg == '--show' {
+	show = true;
+      }
+      else {
+	compopts.push_back(arg);
+      } 
+    }
+  }
+  UpdateLock();
+  BuildProgram(release, show, compopts);
+}
+
+
+proc BuildProgram(release: bool, show: bool, compopts: [?d] string) {
+  if isFile("Mason.lock") {
+
+    // --fast
+    var binLoc = 'debug';
+    if release {
+      binLoc = 'release';
+    }
+
+    // Make Binary Directory
+    makeTargetFiles(binLoc);
+
+    //Install dependencies into .mason/src
+    var toParse = open("Mason.lock", iomode.r);
+    var lockFile = parseToml(toParse);
+    var sourceList = genSourceList(lockFile);
+    getSrcCode(sourceList, show);
+
+    // Compile Program
+    if compileSrc(lockFile, binLoc, show, release, compopts) {
+      writeln("Build Successful\n");
+    }
+    else {
+      writeln("Build Failed");
+    }
+
+    // Close memory
+     delete lockFile;
+     toParse.close();
+  }
+  else {
+    writeln("Cannot build: no Mason.lock found");
+  }
+} 
+
+
+/* Creates the rest of the project structure */
+proc makeTargetFiles(binLoc: string) {
+  if !isDir('target') {
+    mkdir('target'); 
+  }
+  if !isDir('target/' + binLoc) {
+    mkdir('target/' + binLoc);
+    mkdir('target/' + binLoc + '/tests');
+    mkdir('target/'+ binLoc + '/examples');
+    mkdir('target/' + binLoc + '/benches');
+  }
+}
+
+
+/* Compiles the program into the main project src
+   folder. Requires that the main library file be
+   named after the project folder in which it is 
+   contained */
+proc compileSrc(lockFile: Toml, binLoc: string, show: bool, 
+		release: bool, compopts: [?dom] string) : bool {
+  var sourceList = genSourceList(lockFile);
+  var depPath = getMasonHome() + '/.mason/src/';
+  var project = lockFile["root"]["name"].s;
+  var projectPath = 'src/'+ project + '.chpl';
+ 
+  if isFile(projectPath) {
+    var command: string = 'chpl ' + projectPath + ' ' + ' '.join(compopts);
+    if release {
+      command += " --fast";
+    }
+    for dep in sourceList {
+      var depSrc = ' -M '+ depPath + dep(2) + '/src';
+      command += depSrc;
+    }
+
+    // Verbosity control
+    if show {
+      writeln(command);
+    }
+    else {
+      writeln("Compiling "+ project);
+    }
+
+    // compile Program with deps
+    var compilation = runWithStatus(command);
+    if compilation != 0 {
+      return false;
+    }
+    
+    // Confirming File Structure
+    rename(project, 'target/' + binLoc + '/' + project);
```

Comment:
Could this be accomplished with `-o` flag in the compile command? This would also make `--show` output more reproducible for users.